homeassistant/components/stt/__init__.py

```diff
@@ -36,33 +49,171 @@
 
 __all__ = [
     "async_get_provider",
+    "async_get_provider_entity",
     "AudioBitRates",
     "AudioChannels",
     "AudioCodecs",
     "AudioFormats",
     "AudioSampleRates",
     "DOMAIN",
     "Provider",
+    "ProviderEntity",
     "SpeechMetadata",
     "SpeechResult",
     "SpeechResultState",
 ]
 
+_LOGGER = logging.getLogger(__name__)
+
+
+@callback
+def async_get_provider_entity(
+    hass: HomeAssistant, entity_id: str
+) -> ProviderEntity | None:
+    """Return provider entity."""
+    component: EntityComponent[ProviderEntity] = hass.data[DOMAIN]
+
+    return component.get_entity(entity_id)
+
 
 async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up STT."""
+    component = hass.data[DOMAIN] = EntityComponent[ProviderEntity](
+        _LOGGER, DOMAIN, hass
+    )
+
+    # Cancel the retry setup timer on shutdown (temporary workaround)
+    hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STOP,
+        component._async_shutdown,  # pylint: disable=protected-access
+    )
+
     platform_setups = async_setup_legacy(hass, config)
 
     if platform_setups:
         await asyncio.wait([asyncio.create_task(setup) for setup in platform_setups])
 
-    hass.http.register_view(SpeechToTextView(hass.data[DOMAIN]))
+    hass.http.register_view(SpeechToTextView(hass.data[DATA_PROVIDERS]))
     return True
 
 
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent[ProviderEntity] = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent[ProviderEntity] = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+class ProviderEntity(RestoreEntity):
```

Comment:
I'm not sure I like the name of the class, why don't we call it "SpeechToTextEntity"?
Also, why is this a `RestoreEntity`?

Reply:
I can change the name.

Reply:
We discussed the restore feature off line and concluded it's good, since we can't query a device or service for the state on entity addition.